src/mutableAtom/index.ts

```diff
@@ -0,0 +1,155 @@
+import { Getter, Setter, atom } from 'jotai'
+import assert from 'minimalistic-assert'
+import { proxy, snapshot, subscribe } from 'valtio'
+import {
+  Options,
+  PromiseOrValue,
+  ProxyRef,
+  ProxyState,
+  SetCb,
+  SetSelf,
+  Store,
+  Wrapped,
+  WriteFn,
+} from './types'
+
+export function mutableAtom<Value>(
+  value: Value,
+  options: Options<Value> = defaultOptions
+) {
+  const { proxyFn } = { ...defaultOptions, ...options }
+
+  const valueAtom = atom({ value })
+
+  const storeAtom = atom(
+    () => ({ unsubscribe: null, isMounted: false } as Store),
+    (get, set, writeFn: WriteFn) => writeFn(get, set)
+  )
+
+  if (process.env.NODE_ENV !== 'production') {
+    storeAtom.debugPrivate = true
+  }
+
+  storeAtom.onMount = (writeFn) => {
+    // switch to synchronous imperative updates on mount
+    writeFn((get, set) => createProxyState(get, (fn) => fn(set)))
+    return () => writeFn(onAtomUnmount)
+  }
+  /**
+   * unsubscribe on atom unmount
+   */
+  function onAtomUnmount(get: Getter) {
+    get(storeAtom).unsubscribe?.()
+    get(storeAtom).isMounted = false
+  }
+
+  const proxyRefAtom = atom(() => ({ current: null } as ProxyRef<Value>))
+  if (process.env.NODE_ENV !== 'production') {
+    proxyRefAtom.debugPrivate = true
+  }
+
+  /**
+   * sync the proxy state with the atom
+   */
+  function onChange(get: Getter, setCb: SetCb) {
+    return () => {
+      const proxyState = get(proxyRefAtom).current
+      if (proxyState === null) return
+      const { value } = snapshot(proxyState)
+      if (value !== get(valueAtom).value) {
+        setCb((set) => set(valueAtom, { value } as Wrapped<Awaited<Value>>))
+      }
+    }
+  }
+
+  /**
+   * create the proxy state and subscribe to it
+   */
+  function createProxyState(get: Getter, setCb: SetCb) {
+    const proxyRef = get(proxyRefAtom)
+    const store = get(storeAtom)
+    const { value } = get(valueAtom)
+    proxyRef.current ??= proxyFn({ value })
+    proxyRef.current.value = value
+    const unsubscribe = subscribe(proxyRef.current, onChange(get, setCb), true)
+    store.unsubscribe?.()
+    store.unsubscribe = () => {
+      store.unsubscribe = null
+      unsubscribe()
+    }
+    store.isMounted = true
+  }
+
+  /**
+   * return the proxy if it exists, otherwise create and subscribe to it
+   */
+  function ensureProxyState(get: Getter, setCb: SetCb) {
+    const proxyRef = get(proxyRefAtom)
+    const store = get(storeAtom)
+    if (!store.isMounted) createProxyState(get, setCb)
```

Comment:
It seems to me `isMounted` is somewhat misleading.
I would simply do:
```js
if (!proxyRef.current) {
  proxyRef.current = ...
}
```
Now, it may require combining `proxyRefAtom` and `storeAtom`, which I think simplifies the entire code, but I'm not sure.

Reply:
I can add both conditions.

For reasons I don't understand, when I set `proxyRef.current = null` on unmount, then after resubscribing and recreating the proxy on remount, mutating the proxy does not trigger subscribe callback.

*Sequence*
1. unmount:
  a. proxyRef.current = null
  b. unsubscribe()
2. remount:
  a. proxyRef.currrent = proxyFn({ value })
  b. subscribe()
3. mutate proxy:
  a. proxyState.value++
  b. // subscribe does not fire

Solution: don't do `proxyRef.current = null`. I cannot understand why this is.